Add PushToken entity class with validation logic.

hannahtinkler opened the pull request:
### Changes proposed in this Pull Request:

Focus: Woo-driven Push Notifications
Team: Apps Infrastructure
Relates to: AINFRA-1435

Apps Infrastructure are currently working with Kiwi to allow WooCommerce instances to trigger their own push notifications, rather than relying on Jetpack Sync. This change adds the entity that will represent and validate a push token. This will be used in the data store and controller.

**Note:** this is the smallest self-contained change possible, and adding the next file set (data store + tests) will result in the PR being > 1800 lines.

**Key requirements:**
* The push token entity can be instantiate
* The push token entity methods throw if invalid data is provided
* The push token entity can determine if it has the correct data to perform a CRUD action

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:
There are no functional changes to test in this PR, but to ensure no side effects, you can:
* Check out this branch to your WooCommerce install
* Ensure your WooCommerce install functions as expected, e.g.:
  * Add a product to cart
  * View WooCommerce settings in WP Admin
  * View WooCommerce orders in WP Admin
* Make an authenticated request to the API to ensure this works as expected, e.g.:
  * https://YOUR_URL/wp-json/wc-admin/features
  * https://YOUR_URL/wp-json/wc/v3/settings
  * https://YOUR_URL/wp-json/wc/v3/customers

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Adds internal library foundations which will eventually be used to send push notifications to user devices. This will support WooCommerce in directly triggering push notifications, instead of relying on Jetpack Sync.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->
This is an internal, non-user facing change.

</details>

hannahtinkler committed:
Add PushToken entity class with validation logic.

- Implement PushToken entity with getters/setters for all properties
- Add validation for platforms, origins, token length, and IDs
- Include CRUD capability checks (can_be_created, can_be_updated, etc.)
- Handle browser platform exception for device UUID requirement

Konamiman commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -0,0 +1,349 @@
+	 *
+	 * @since 10.4.0
+	 */
+	public function can_be_created(): bool {
```
This method is almost identical to `can_be_updated`. I think we could:

1. Create a `private can_be_create_or_updated(bool $id_must_exist): bool` method with the logic.
2. `can_be_created(): { return $this->can_be_create_or_updated(false) }`, and the opposite for the other method.

hannahtinkler replied:
I've implemented a version of this that removes the duplication, WDYT? It reads a little clearer to me with the ID check in the `can_be_*` methods still

Konamiman replied:
Yes, good idea, an id check + method call is as clear as a single method call if not more.

Konamiman commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -0,0 +1,349 @@
+ *
+ * @since 10.4.0
+ */
+class PushToken {
```
If I understand correctly, this is not the usual singleton class but a "DTO like" class intended to be created directly with `new`; also various `set_*` methods need to be invoked before the class is actually usable.

If that's the case, I suggest adding a public static `get_new_instance(?$int id=null, ?$int user_id=null, ?$string token=null, ?string $device_uuid=null, ?string $platform=null, ?string $origin=null)` method that `new`s a new instance and invokes all the `set_*` methods. This would also make unit tests a bit shorter.

hannahtinkler replied:
Although we can use an instance of PushToken without the full data set (e.g. by providing the ID and passing it to `read` to get the rest), it isn't a really a valid instance of a push token unless it has all of the required data. It feels a bit strange to have a `get_new_instance` method that could return an 'invalid' instance 🤔 What's the main driver behind the suggestion? I'm wondering if there's another way to achieve it...I did try out a `from_array` method:

```
$push_token = PushToken::from_array(
	array(
		'user_id'     => 1,
		'token'       => 'test_token',
		'device_uuid' => 'test-device-uuid',
		'platform'    => PushToken::PLATFORM_IOS,
		'origin'      => PushToken::ORIGIN_WOOCOMMERCE_IOS,
	)
);
```
...but it uses more lines so didn't seem like a viable alternative since you mentioned making the tests shorter 😆

Konamiman replied:
Think of `get_new_instance` as just an utility method that provides you a one-liner to create an instance with all the data you want to specify upfront: whether the instance generated by such method is valid or not is still determined by the internal logic of the class (the `get_new_instance` name doesn't imply that the instance will be valid). To be extra sure, the docblock of that method could explicity include something like "using this method doesn't imply that the instance obtained is valid".

And indeed, it's cleaner and less error-prone to provide the required data in dedicated arguments rather than an array, provided that there aren't too much of those (six is still reasonable I think).

hannahtinkler replied:
I'm happy to add it if you feel strongly 🙂 I think that's all the amends taken care of if you want to take another look?

Konamiman commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -0,0 +1,349 @@
+		}
+
+		if ( strlen( $token ) > self::MAX_TOKEN_LENGTH ) {
+			throw new InvalidArgumentException( 'Token exceeds maximum length.' );
```
Suggestion: include the maximum length in the exception message.

Konamiman reviewed (commented):
Overall it looks good, I left some suggestions for minor improvements. Also for completeness, we could add a test for the trimming of the UUID, like this:

```php
  /**
   * @testdox Tests device UUID trims whitespace.
   */
  public function test_it_trims_whitespace_from_device_uuid() {
      $push_token = new PushToken();
      $push_token->set_device_uuid( '  test-uuid  ' );

      $this->assertEquals( 'test_uuid', $push_token->get_device_uuid() );
  }
```

Konamiman reviewed (changes requested):
Overall it looks good, I left some suggestions for minor improvements. Also for completeness, we could add a test for the trimming of the UUID, like this:

```php
  /**
   * @testdox Tests device UUID trims whitespace.
   */
  public function test_it_trims_whitespace_from_device_uuid() {
      $push_token = new PushToken();
      $push_token->set_device_uuid( '  test-uuid  ' );

      $this->assertEquals( 'test_uuid', $push_token->get_device_uuid() );
  }
```

hannahtinkler committed:
Improve PushToken validation and error messages.

- Add specific token length in error message for clarity.
- Extract common validation logic to private has_required_parameters method.

hannahtinkler committed:
Add get_new_instance utility method to PushToken class.

- Add static factory method to create PushToken instances with properties.
- Refactor existing tests to use get_new_instance for cleaner code.

hannahtinkler committed:
Updates code comment typo.

hannahtinkler committed:
Updates typo in code comment.
